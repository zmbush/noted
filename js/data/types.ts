// This file is auto-generated by tools/generate_types.js
// Do not modify this file directly!

export interface ErrorData {
  code: number;
  column_name?: string | null;
  constraint_name?: string | null;
  details?: string | null;
  error: string;
  hint?: string | null;
  table_name?: string | null;
}

export interface NewNote {
  body: string;
  parent_note_id?: number | null;
  title: string;
}

export interface NewUserRequest {
  email: string;
  name: string;
  password: string;
}

export interface NoteWithTags {
  archived: boolean;
  body: string;
  created_at: string;
  id: number;
  parent_note_id: number;
  pinned: boolean;
  tags: string[];
  title: string;
  updated_at: string;
  user_id: number;
}

export interface SignIn {
  email: string;
  password: string;
}

export interface UpdateNote {
  archived?: boolean | null;
  body?: string | null;
  parent_note_id?: number | null;
  pinned?: boolean | null;
  title?: string | null;
}

export interface User {
  created_at: string;
  email: string;
  id: number;
  name: string;
  updated_at: string;
}
