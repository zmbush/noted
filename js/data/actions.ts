// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import { NoteData, ErrorData, UserData } from 'data/types';
import axios from 'axios';
import { Dispatch } from 'redux';

export enum NotedEvent {
  NotesFetchStart = 'NOTES_FETCH_START',
  NotesFetched = 'NOTES_FETCHED',
  NotesUpdateNote = 'NOTES_UPDATE_NOTE',
  NotesDeleteNote = 'NOTES_DELETE_NOTE',

  ApiError = 'API_ERROR',

  UserSignedIn = 'USER_SIGNED_IN',
  UserSignedOut = 'USER_SIGNED_OUT',
}

export function notesFetchStart() {
  return { type: NotedEvent.NotesFetchStart };
}

export function notesFetched(notes: NoteData[]) {
  return { type: NotedEvent.NotesFetched, notes };
}

export function updateNote(note: NoteData) {
  return { type: NotedEvent.NotesUpdateNote, note };
}

export function deleteNote(id: number) {
  return { type: NotedEvent.NotesDeleteNote, id };
}

export function apiError(error: ErrorData) {
  return { type: NotedEvent.ApiError, error };
}

export function logIn(user: UserData) {
  return { type: NotedEvent.UserSignedIn, user };
}

export function logOut() {
  return { type: NotedEvent.UserSignedOut };
}

export async function fetchData(dispatch: Dispatch) {
  dispatch(notesFetchStart());
  try {
    dispatch(notesFetched((await axios.get('/api/secure/notes')).data));
  } catch (e) {
    if (e.response) {
      dispatch(apiError(e.response.data));
    }
  }
}
