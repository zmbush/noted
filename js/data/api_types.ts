// This file is auto-generated by tools/generate_api_types.js
// Do not modify this file directly!

export interface NoteWithTags {
  archived: boolean;
  body: string;
  created_at: string;
  id: number;
  parent_note_id: number;
  pinned: boolean;
  tags: string[];
  title: string;
  updated_at: string;
  user_id: number;
  [k: string]: unknown;
}

export interface User {
  created_at: string;
  email: string;
  id: number;
  name: string;
  updated_at: string;
  [k: string]: unknown;
}
