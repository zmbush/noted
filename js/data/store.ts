// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { AnyAction, combineReducers, configureStore } from '@reduxjs/toolkit';

import notes from 'data/notes/slice';
import ui from 'data/ui/slice';
import user from 'data/user/slice';

export const appReducer = combineReducers({ user, notes, ui });

export const rootReducer = (state: ReturnType<typeof appReducer>, action: AnyAction) => {
  if (action.type === 'RESET') {
    return appReducer(undefined, action);
  }
  return appReducer(state, action);
};
export const store = configureStore({
  reducer: rootReducer,
  devTools: true,
});

export type AppState = ReturnType<typeof store.getState>;
export type AppDispatch = typeof store.dispatch;
