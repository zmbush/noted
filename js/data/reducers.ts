// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import { combineReducers } from 'redux';
import { NotedEvent } from 'data/actions';

function notes(
  state = new Map<number, any>(),
  action: { type: NotedEvent; notes?: any[]; note?: any }
) {
  switch (action.type) {
    case NotedEvent.NOTES_FETCHED: {
      let data = new Map();
      for (let note of action.notes) {
        data.set(note.id, note);
      }
      return data;
    }
    case NotedEvent.NOTES_UPDATE_NOTE: {
      let data = new Map(state);
      data.set(action.note.id, action.note);
      return data;
    }
    default:
      return state;
  }
}

export default combineReducers({ notes });
