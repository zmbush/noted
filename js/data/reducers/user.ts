// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import { NotedEvent } from 'data/actions';
import { UserData, ErrorData } from 'data/types';
import update from 'immutability-helper';

const initialState = { is_signed_in: false, user: null as (null | UserData) };

type State = typeof initialState;

export default function user(
  state = initialState,
  action: {
    type?: NotedEvent;
    error?: ErrorData;
    is_signed_in?: boolean;
    user?: UserData;
  }
): State {
  switch (action.type) {
    case NotedEvent.ApiError: {
      if (action.error.code == 401) {
        return initialState;
      }
    }
    case NotedEvent.UserSignedIn: {
      return update(state, {
        is_signed_in: { $set: true },
        user: { $set: action.user },
      });
    }
    case NotedEvent.UserSignedOut: {
      return initialState;
    }
  }
  return state;
}
