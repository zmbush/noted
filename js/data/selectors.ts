// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import { createSelector } from 'reselect';
import createCachedSelector from 're-reselect';
import { NoteData, AppState } from 'data/types';

const getNotes = (state: AppState) => state.notes;
const getNoteId = (_: any, props: { note_id: number }) => props.note_id;

const listNotes = createSelector(
  getNotes,
  (notes: Map<number, NoteData>) => Array.from(notes.values()),
);

const validParent = (note: NoteData) => !!note.parent_note_id;

const listAllTitles = createSelector(
  listNotes,
  notes => notes.map(note => [note.id, note.title]),
);

const listAllTitleParts = createSelector(
  listAllTitles,
  (titleList: [number, string][]) =>
    titleList.flatMap(([id, title]) => {
      const titles = [[title, id]];
      title.split(' ').forEach(titlePart => {
        if (titlePart.length > 3) {
          titles.push([titlePart, id]);
        }
      });
      return titles;
    }),
);

export const getLinkIds = createSelector(
  listAllTitleParts,
  (titleParts: [string, number][]) =>
    titleParts.reduce((titles, [titlePart, id]) => {
      if (titles.has(titlePart)) {
        titles.get(titlePart).add(id);
      } else {
        titles.set(titlePart, new Set([id]));
      }
      return titles;
    }, new Map<string, Set<number>>()),
);

export type LinkIdMap = ReturnType<typeof getLinkIds>;

export const getTopLevelNotes = createSelector(
  listNotes,
  notes => {
    const topLevel = new Map();
    notes.forEach(note => {
      if (!validParent(note)) {
        topLevel.set(note.id, note);
      }
    });
    return topLevel;
  },
);

export const getSubnotes = createCachedSelector(listNotes, getNoteId, (notes, noteId) => {
  const subnotes = new Map();

  [...notes.values()].forEach(note => {
    if (note.parent_note_id === noteId) {
      subnotes.set(note.id, note);
    }
  });

  return subnotes;
})((state, props: { note_id: number }) => props.note_id);

export type SubnoteMap = ReturnType<typeof getSubnotes>;

const calculateMergedNote = (note: NoteData, notes: Map<number, NoteData>): NoteData => {
  const newNote = { ...note };
  const subnotes = getSubnotes({ notes }, { note_id: note.id });

  // eslint-disable-next-line no-restricted-syntax
  [...subnotes.values()].forEach(subnote => {
    const mergedNote = calculateMergedNote(subnote, notes);
    newNote.title = `${newNote.title} ${mergedNote.title}`;
    newNote.tags = Array.from(new Set([...newNote.tags, ...mergedNote.tags]));
    newNote.body = `${newNote.body} ${mergedNote.body}`;
  });

  return newNote;
};

export const getSearchIndex = createSelector(
  getNotes,
  allNotes => {
    const result = new Map();
    [...allNotes.values()].forEach(note => {
      result.set(note.id, calculateMergedNote(note, allNotes));
    });
    return result;
  },
);

export const getFilteredSearchIndex = createCachedSelector(
  getSearchIndex,
  getNoteId,
  (notes, noteId) => {
    const subnotes = new Map();
    if (noteId == null) {
      // eslint-disable-next-line no-param-reassign
      noteId = 0;
    }

    [...notes.values()].forEach(note => {
      if (note.parent_note_id === noteId) {
        subnotes.set(note.id, note);
      }
    });

    return subnotes;
  },
)((state, props: { note_id: number }) => (props.note_id == null ? -1 : props.note_id));

const mostRecent = (a: string, b: string) => (a > b ? a : b);

export const getSortedNoteIds = createSelector(
  getNotes,
  allNotes => {
    const map = new Map();

    [...allNotes.values()].forEach(note => {
      map.set(note.id, note.updated_at);
    });

    [...allNotes.values()].forEach(note => {
      let curNote = note;
      while (validParent(curNote)) {
        map.set(note.parent_note_id, mostRecent(note.updated_at, map.get(note.parent_note_id)));
        curNote = allNotes.get(note.parent_note_id);
      }
    });

    return Array.from(map.entries())
      .sort(([aid, a]: [number, string], [bid, b]: [number, string]) => {
        const na = allNotes.get(aid);
        const nb = allNotes.get(bid);

        if (na.pinned && !nb.pinned) {
          return -1;
        }
        if (!na.pinned && nb.pinned) {
          return 1;
        }

        if (na.archived && !nb.archived) {
          return 1;
        }
        if (!na.archived && nb.archived) {
          return -1;
        }

        if (a > b) {
          return -1;
        }
        if (a < b) {
          return 1;
        }
        return 0;
      })
      .map(([id, _]) => id);
  },
);
