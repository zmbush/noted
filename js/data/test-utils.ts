// Copyright 2021 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.
//
import { rootReducer } from 'data/store';
import { NoteWithTags } from 'data/types';

import { getNotes } from './notes/api';

export const emptyTestNote: NoteWithTags = {
  id: -1,
  user_id: -1,
  title: 'The Note',
  body: 'The Body',
  tags: [],
  archived: false,
  pinned: false,
  parent_note_id: 0,
  created_at: '',
  updated_at: '',
};

let id = 0;
export const makeTestNote = (
  overrides: Partial<NoteWithTags> = {},
  parent: { id: number } = null,
): NoteWithTags => {
  id += 1;
  const thisId = overrides.id || id;
  const defaultFields: Partial<NoteWithTags> = {
    id: thisId,
    title: `Note ${thisId}`,
  };
  const parentOverride: Partial<NoteWithTags> = {};
  if (parent) {
    parentOverride.parent_note_id = parent.id;
  }
  return { ...emptyTestNote, ...defaultFields, ...overrides, ...parentOverride };
};

export const makeTestState = () =>
  rootReducer(
    undefined,
    getNotes.fulfilled(
      [makeTestNote(), makeTestNote({ id: 4 }), makeTestNote({ id: 2 })],
      '',
      undefined,
      undefined,
    ),
  );
