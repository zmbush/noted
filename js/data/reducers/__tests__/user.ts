// Copyright 2019 Zachary Bush.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
// option. This file may not be copied, modified, or distributed
// except according to those terms.

import user from '../user';
import { NotedEvent } from 'data/actions';
import { NoteData } from 'data/types';

describe('reducers::user()', () => {
  let id = 1;
  const getInitial = () => user(undefined, {});

  test('returns initial state', () => {
    expect(getInitial()).toEqual({ is_signed_in: false, user: null });
  });

  test('responds to events', () => {
    let state = getInitial();
    let u = {
      id: 1,
      name: 'Test User',
      email: 'test@example.com',
      created_at: '',
      updated_at: '',
    };
    state = user(state, { type: NotedEvent.UserSignedIn, user: u });

    expect(state).toMatchSnapshot();
    expect(user(state, { type: NotedEvent.UserSignedOut })).toMatchSnapshot();
    expect(
      user(state, {
        type: NotedEvent.ApiError,
        error: { code: 401, error: '' },
      })
    ).toMatchSnapshot();
  });
});
