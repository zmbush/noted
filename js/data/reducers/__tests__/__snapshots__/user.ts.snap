// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers::user() responds to events 1`] = `
Object {
  "is_signed_in": true,
  "user": Object {
    "created_at": "",
    "email": "test@example.com",
    "id": 1,
    "name": "Test User",
    "updated_at": "",
  },
}
`;

exports[`reducers::user() responds to events 2`] = `
Object {
  "is_signed_in": false,
  "user": null,
}
`;

exports[`reducers::user() responds to events 3`] = `
Object {
  "is_signed_in": false,
  "user": null,
}
`;
